src/main/kotlin/com/example/delivery/service/OrderService.kt

```diff
@@ -2,54 +2,86 @@ package com.example.delivery.service
+import org.bson.types.ObjectId
 import org.springframework.stereotype.Service
 
 @Service
 class OrderService(
     private val orderRepository: OrderRepository,
     private val userRepository: UserRepository,
-    private val productReserveService: ProductReserveService,
     private val productRepository: ProductRepository,
 ) {
 
     @LogInvoke
-    fun getById(id: String): DomainOrder {
+    fun getById(id: String): DomainOrderWithProduct {
         return orderRepository.findById(id)?.toDomain()
             ?: throw NotFoundException("Order with id $id doesn't exists")
     }
 
+    @SuppressWarnings("ThrowsCount")
     fun add(createOrderDTO: CreateOrderDTO): DomainOrder {
         val user = userRepository.findById(createOrderDTO.userId)
             ?: throw NotFoundException("User with id ${createOrderDTO.userId} doesn't exist")
 
-        val products = createOrderDTO.items.mapKeys {
-            productRepository.findById(it.key)?.toDomain()
-                ?: throw NotFoundException("Product with id ${it.key} doesn't exist")
-        }.map { ProductReserveService.AccountedProduct(it.key, it.value) }
+        val products = orderRepository.fetchProducts(createOrderDTO.items.map { it.productId }).map { it.toDomain() }
 
-        productReserveService.reserveProducts(products)
+        if (products.size != createOrderDTO.items.size) {
+            val missingProductId = createOrderDTO.items
+                .map { it.productId }
+                .first { itemId -> !products.map { it.id.toString() }.contains(itemId) }
+            throw NotFoundException("Product with id $missingProductId")
+        }
 
-        productRepository.updateProductsAmount(products.associate { it.product.id.toString() to -it.amount })
+        createOrderDTO.items.forEach { orderItem ->
+            val product = products.first { orderItem.productId == it.id.toString() }
+            if (orderItem.amount > product.amountAvailable) {
+                throw ProductAmountException(
+                    "Insufficient stock for product ${product.name}. " +
+                        "Available: ${product.amountAvailable}, " +
+                        "Requested: ${orderItem.amount}"
+                )
+            }
+        }
 
-        val totalPrice = productReserveService.calculateTotalPrice(products)
+        productRepository.updateProductsAmount(
+            createOrderDTO.items.map {
+                MongoOrder.MongoOrderItem(
+                    ObjectId(it.productId),
+                    null,
+                    it.amount
+                )
+            }
+        )
+
+        val mongoOrderItems = createOrderDTO.items.map { item ->
+            MongoOrder.MongoOrderItem(
+                ObjectId(item.productId),
+                products.first { it.id.toString() == item.productId }.price,
+                item.amount
+            )
+        }
+
+        productRepository.updateProductsAmount(mongoOrderItems)
```

Comment:
why do you call updateProductsAmount 2 times?

Reply:
🤦